talker_tests/uts/test_client.py

```diff
@@ -0,0 +1,118 @@
+                                pid: str = '1234', delay: float = None) -> None:
+        if delay:
+            sleep(delay)
+        ack = time()
+        self.redis.rpush('result-{}-stdout'.format(job_id), stdout_val)
+        self.redis.rpush('result-{}-stderr'.format(job_id), stderr_val)
+        self.redis.rpush('result-{}-ack'.format(job_id), ack)
+        self.redis.rpush('result-{}-retcode'.format(job_id), retcode)
+        self.redis.rpush('result-{}-pid'.format(job_id), pid)
+
+    def get_command_from_redis(self) -> dict:
+        cmd_key = 'commands-{}'.format(self.host_id)
+        ret = self.redis.blpop([cmd_key], timeout=1)
+        _, job_data_raw = ret
+        result = json.loads(job_data_raw)
+        return result
+
+    def test_run_command_is_written_to_redis(self) -> None:
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'echo hello')
+        sleep(0.1)  # write to redis must be completed before we pop from redis
+        result = self.get_command_from_redis()
+        expected_value = {'id': cmd.job_id,
+                          'cmd': ['bash', '-ce', 'echo hello'],
+                          'timeout': 3600.0}
+        self.assertEqual(result, expected_value)
+
+    def test_command_result_from_agent(self) -> None:
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'echo hello')
+        self.write_response_to_redis(cmd.job_id, 'hello\n')
+        res = cmd.result()
+        self.assertEqual(res, 'hello\n')
+
+    def test_client_wait_command(self) -> None:
+        delay = 0.5
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'sleep {}'.format(delay))
+        before = time()
+        Thread(target=self.write_response_to_redis, kwargs={'job_id': cmd.job_id, 'delay': delay}).start()
+        ret = cmd.wait()
+        after = time()
+        self.assertEqual(ret, 0)
+        self.assertTrue(delay - delay * 0.1 <= after - before <= delay + delay * 0.1)
+
+    def test_timeout_error(self) -> None:
+        delay = 5
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'sleep {}'.format(delay), timeout=delay * 0.0001,
+                              server_timeout=False)
+        Thread(target=self.write_response_to_redis, kwargs={'job_id': cmd.job_id, 'delay': delay},
+               daemon=True).start()
+        try:
+            _ = cmd.wait()
+            raise Exception('ClientCommandTimeoutError was not raised')
+        except ClientCommandTimeoutError:
+            pass
+
+    def test_command_fail(self) -> None:
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'exit 2')
+        self.write_response_to_redis(cmd.job_id, retcode=2)
+        try:
+            cmd.result()
+            raise Exception('CommandExecutionError was not raised')
+        except CommandExecutionError:
+            pass
+
+    def test_ignoring_command_fail(self) -> None:
+        cmd = self.client.run(self.host_id, 'bash', '-ce', 'exit 2', raise_on_failure=False)
```

Comment:
another way to ignore this failure is to pass a `retcode`. you can test this here too.
```
cmd = self.client.run(..., retcode=(2, ))
```
`retcode` is a tuple of acceptable return codes from the command.

Reply:
I assume you mean something like `cmd.retcode = (2,)` because self.client.run doesn't accept retcode.

Reply:
my bad. it isn't available in `Talker.run` you can forget about it